docs/source/notes/modules.rst

```diff
@@ -0,0 +1,433 @@
+     def forward(self, x):
+       self.mean = self.momentum * self.mean + (1.0 - self.momentum) * x
+       return self.mean
+
+Now, the current value of the running mean is considered part of the module's ``state_dict``
+and will be saved as part of the module when the module is serialized.
+
+.. code-block:: python
+
+   m = RunningMean(4)
+   for _ in range(10):
+     input = torch.randn(4)
+     m(input)
+
+   print(m.state_dict())
+   : OrderedDict([('mean', tensor([ 0.1041, -0.1113, -0.0647,  0.1515]))]))
+
+   # Serialized form will contain the 'mean' tensor
+   torch.save(m.state_dict(), 'mean.pt')
+
+   m_loaded = RunningMean(4)
+   m_loaded.load_state_dict(torch.load('mean.pt'))
+   assert(torch.all(m.mean == m_loaded.mean))
+
+As mentioned previously, buffers can be left out of the module's ``state_dict`` by marking them as non-persistent:
+
+.. code-block:: python
+
+   self.register_buffer('unserialized_thing', torch.randn(5), persistent=False)
+
+Both persistent and non-persistent buffers are included in model-wide device / dtype changes:
+
+.. code-block:: python
+
+   # Moves all module parameters and buffers to the specified device / dtype
+   m.to(device='cuda', dtype=torch.float64)
+
+Buffers of a module can be iterated over using :func:`~torch.nn.Module.buffers` or
+:func:`~torch.nn.Module.named_buffers`.
+
+For more information, check out:
+
+* Saving and loading: https://pytorch.org/tutorials/beginner/saving_loading_models.html
+* Serialization semantics: https://pytorch.org/docs/master/notes/serialization.html
+* What is a state dict? https://pytorch.org/tutorials/recipes/recipes/what_is_state_dict.html
+
+Module Hooks
+------------
+
+To ensure that simple modifications can be done easily, even for modules that you haven't written yourself,
+PyTorch provides two types of hooks for modules:
+
+* **Forward hooks** are called during the forward pass. They can be installed with
+  :func:`~torch.nn.Module.register_forward_pre_hook` and :func:`~torch.nn.Module.register_forward_hook`.
+  These hooks will be called respectively just before the forward function is called and just after it is called.
+* **Backward hooks** are called during the backward pass. They can be installed with
+  :func:`~torch.nn.Module.register_full_backward_hook`. These hooks will be called when the backward for this
+  Module has been computed and will allow the user to access the gradients for both the inputs and outputs.
+
+All hooks allow the user to return an updated value that will be used throughout the remaining computation.
+Thus, these hooks can be used to either execute arbitrary code along the regular module forward/backward or
+modify some inputs/outputs without having to change the module's ``forward`` function.
+
+PyTorch also provides a way to register these hooks on all existing modules at once by using the
+``register_module_XXX`` version. This can be very useful when building profiling or visualization tools that
```

Comment:
I'm not sure what "version" means here.

Reply:
I rewrote this a bit to mention the global hook registration functions in the respective forward / backward sections.